### PR TITLE
Add metrics to Kafka exporter

### PR DESCRIPTION
**Pull request by dshomoye**
Add metrics otlp exporter to kafka (previously only supported traces).
- add tests for kafka metrics exporter.

**Description:** 
- The kafka exporter now supports both traces and metrics
- Metrics is encoded as otlp proto by default (same as Traces, also the only encoding currently implemented)
- Update the configuration of the kafka exporter to accomodate traces and metrics
  - `encoding` key specifies encoding for either metrics or traces.
  - Add default topic for metrics as `otlp_metrics`.

**Testing:** 
- Unit tests added for metrics exporter (97.1% package cov)
- Was able to deploy the demo set up(`/examples/demo`). Successfully received metrics and traces in a kafka consumer

**Documentation:**

Updated kafka exporter README to state new defaults.
Updated CHANGELOG to reflect enhancement.

---

### CODE REVIEW

File: exporter/kafkaexporter/README.md [hunk 1]
```diff
@@ -9,11 +9,13 @@ The following settings are required:
 
 The following settings can be optionally configured:
 - `brokers` (default = localhost:9092): The list of kafka brokers
-- `topic` (default = otlp_spans): The name of the kafka topic to export to
-- `encoding` (default = otlp_proto): The encoding of the payload sent to kafka. Available encodings:
-  - `otlp_proto`: the payload is serialized to `ExportTraceServiceRequest`.
+- `topic` (default = otlp): The name of the kafka topic to export to
+- `traces_encoding` (default = otlp_proto): The encoding of the traces sent to kafka. All available traces encodings:
```
Comment (tigrannajaryan) [inline exporter/kafkaexporter/README.md:13]: This is a breaking change. Is it needed? What's the use case when traces and metrics need to use different encodings? Even if there is a such use case one can have 2 different exporters configured with different encodings in that case.

Can we keep the `encoding` key? Obviously settings that only work for traces will be invalid for a metric exporter (e.g. `jaeger_proto` should not be used for an exporter attached to metric pipeline).

Comment (dshomoye) [reply exporter/kafkaexporter/README.md:13]: Right, this *is* a breaking change, potential confusion between valid encoding based on metrics/traces is why it was split up this way. I will revert this to use only the `encoding` key and update the instructions to indicate which encoding can be used for which context

**Comment by dshomoye** [comment]
Adding a comment here to bring this PR up again, @tigrannajaryan just want to confirm there is nothing else needed on my end for this to move forward?

**Review by asuresh4 — COMMENTED** [review]
@dshomoye - apologies for the delay in review. Mostly LGTM, I left a few comments. 

@tigrannajaryan @bogdandrutu would be great if you or anyone more familiar with the kafka exporter could also take a look.

File: exporter/kafkaexporter/kafka_exporter.go [hunk 2]
```diff
@@ -28,21 +28,22 @@ import (
 
 var errUnrecognizedEncoding = fmt.Errorf("unrecognized encoding")
 
-// kafkaProducer uses sarama to produce messages to Kafka.
-type kafkaProducer struct {
+// kafkaTracesProducer uses sarama to produce messages to Kafka.
+type kafkaTracesProducer struct {
```
[the remark this answers is not in the document]
Comment (dshomoye) [reply exporter/kafkaexporter/kafka_exporter.go:32]: Good point, I'll rearrange those and update for the other 2 comments as well. 👍

File: exporter/kafkaexporter/README.md [hunk 3]
```diff
@@ -9,11 +9,12 @@ The following settings are required:
 
 The following settings can be optionally configured:
 - `brokers` (default = localhost:9092): The list of kafka brokers
-- `topic` (default = otlp_spans): The name of the kafka topic to export to
-- `encoding` (default = otlp_proto): The encoding of the payload sent to kafka. Available encodings:
-  - `otlp_proto`: the payload is serialized to `ExportTraceServiceRequest`.
-  - `jaeger_proto`: the payload is serialized to a single Jaeger proto `Span`.
-  - `jaeger_json`: the payload is serialized to a single Jaeger JSON Span using `jsonpb`.
+- `topic` (default = otlp): The name of the kafka topic to export to
```
Comment (tigrannajaryan) [inline exporter/kafkaexporter/README.md:12]: This is a breaking change. Make sure to mention in the CHANGELOG.

File: exporter/kafkaexporter/factory.go [hunk 4]
```diff
@@ -108,3 +111,22 @@ func (f *kafkaExporterFactory) createTraceExporter(
...
+) (component.MetricsExporter, error) {
+	oCfg := cfg.(*Config)
+	exp, err := newMetricsExporter(*oCfg, params, f.metricsMarshallers)
+	if err != nil {
+		return nil, err
+	}
+	return exporterhelper.NewMetricsExporter(
+		cfg,
+		exp.metricsDataPusher,
+		exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0}),
```
Comment (tigrannajaryan) [inline exporter/kafkaexporter/factory.go:128]: Copy the comment from lines 107-108 here since it applies here too.

File: exporter/kafkaexporter/otlp_marshaller_test.go [hunk 5]
```diff
@@ -21,24 +21,39 @@ import (
...
 
-	m := otlpProtoMarshaller{}
+	m := otlpTracesPbMarshaller{}
 	assert.Equal(t, "otlp_proto", m.Encoding())
 	messages, err := m.Marshal(td)
 	require.NoError(t, err)
 	assert.Equal(t, []Message{{Value: expected}}, messages)
 }
+
+func TestOTLPMetricsPbJSONMarshaller(t *testing.T) {
```
Comment (tigrannajaryan) [inline exporter/kafkaexporter/otlp_marshaller_test.go:45]: Why does this have "JSON" in the name? Does it do anything related to JSON? This is binary Protobuf marshaler test, right?

Comment (dshomoye) [reply exporter/kafkaexporter/otlp_marshaller_test.go:45]: Yes it is only a Proto test, this is a typo, renaming it now.

File: exporter/kafkaexporter/README.md (hunk 3 above)
Comment (dshomoye) [reply exporter/kafkaexporter/README.md:12]: Looking at the current CHANGELOG file, does this breaking have to be associated with a specific version or filed under "Unreleased"

**Comment by tigrannajaryan** [comment]
@pavolloffay please review, since you are original author of the component.

**Comment by dshomoye** [comment]
Adding another comment since this has been almost a week, making sure this doesn't slip under the radar again. 

> @pavolloffay please review, since you are original author of the component.

File: CHANGELOG.md [hunk 6]
```diff
@@ -12,6 +12,11 @@
 - Rename component.TraceReceiver to component.TracesReceiver #1975
 - Rename component.TraceProcessor to component.TracesProcessor #1976
 - Rename component.TraceExporter to component.TracesExporter #1975
+- `kafka` exporter: Rename default topic to `otlp` from `otlp_spans`
```
Comment (pavolloffay) [inline CHANGELOG.md:15]: heads up @jpkrohling @objectiser

**Comment by pavolloffay** [comment]
> - `kafka` exporter: Rename default topic to `otlp` from `otlp_spans`

What is the motivation to change the topic name? If you proceed with rename the receiver should be updated as well.

**Comment by dshomoye** [comment]
> > * `kafka` exporter: Rename default topic to `otlp` from `otlp_spans`
> 
> What is the motivation to change the topic name? If you proceed with rename the receiver should be updated as well.

Primarily because `*span` is a trace-specific name; sending metrics to a `*span` topic seems a bit out of place.
As for updating the receiver, there hasn't been any changes for metrics support there so it wasn't updated. Is changing the topic for consistency or is there some dependency between exporter/receiver here? (my impression is that there isn't)

**Comment by jpkrohling** [comment]
> Is changing the topic for consistency or is there some dependency between exporter/receiver here?

I believe we are still allowed to make breaking changes, but people might be using this exporter already and rely on the previously published name with their own consumers. Ideally, changes like these would happen in two phases:

1. announce the change, deprecating the previous value. Log a info-level message telling that this will change (unless it's explicitly set)
1. one release further, change the level to warn (unless it's explicitly set)
1. two releases further, make the switch

**Comment by dshomoye** [comment]
> people might be using this exporter already and rely on the previously published name with their own consumers. 
Understood, in that case, I will revert this breaking change and add the log message for now.

**Comment by pavolloffay** [comment]
@dshomoye re the topic name, isn't it a better practice to __not send__ various types to the same topic? If so we should document this and use different default per metrics and traces.

**Comment by dshomoye** [comment]
> @dshomoye re the topic name, isn't it a better practice to **not send** various types to the same topic? If so we should document this and use different default per metrics and traces.
> 
> Using the same topic name for both types will break the current kafka receiver, it will fail to unmarshall traces.

You're absolutely right! Mixing different data types in the same topic is not good practice - even though the user can override for trace/metric. 
Still, this actually resolves the breaking change issue. The default topic for traces will stay as `otlp_spans` and a new default topic `otlp_metrics` will be added for metrics. No deprecation required.

**Review by bogdandrutu — APPROVED** [review]
For me this looks good, @pavolloffay @tigrannajaryan  is it ok to merge this?

File: service/internal/resources.go [hunk 7]
```diff
@@ -227,7 +227,7 @@ var _escData = map[string]*_escFile{
 		name:    "component_header.html",
 		local:   "templates/component_header.html",
 		size:    156,
-		modtime: 1594178791,
+		modtime: 1602111360,
```
Comment (tigrannajaryan) [inline service/internal/resources.go:230]: This file probably should not be changed in this PR.

Comment (dshomoye) [reply service/internal/resources.go:230]: I'll remove it now

File: CHANGELOG.md [hunk 8]
```diff
@@ -18,6 +22,9 @@
 - Rename component.TraceReceiver to component.TracesReceiver #1975
 - Rename component.TraceProcessor to component.TracesProcessor #1976
 - Rename component.TraceExporter to component.TracesExporter #1975
+
+## 💡 Enhancements 💡
```
Comment (pavolloffay) [inline CHANGELOG.md:26]: Why did you add this? The enhancements section is below

File: CHANGELOG.md [hunk 9]
```diff
@@ -83,6 +90,7 @@
 - Fix panic in otlp traces to zipkin (#1963)
 - Fix OTLP/HTTP receiver's path to be /v1/traces (#1979)
 
+
```
Comment (pavolloffay) [inline CHANGELOG.md:93]: nit: keep the changes minimal and only related to your PR

File: exporter/kafkaexporter/config.go [hunk 10]
```diff
@@ -32,9 +32,10 @@ type Config struct {
 	Brokers []string `mapstructure:"brokers"`
 	// Kafka protocol version
 	ProtocolVersion string `mapstructure:"protocol_version"`
-	// The name of the kafka topic to export to (default "otlp_spans")
+	// The name of the kafka topic to export to (default "otlp")
```
Comment (pavolloffay) [inline exporter/kafkaexporter/config.go:35]: the comment is wrong

File: exporter/kafkaexporter/factory.go [hunk 11]
```diff
@@ -72,8 +75,10 @@ func createDefaultConfig() configmodels.Exporter {
 		RetrySettings:   exporterhelper.CreateDefaultRetrySettings(),
 		QueueSettings:   exporterhelper.CreateDefaultQueueSettings(),
 		Brokers:         []string{defaultBroker},
-		Topic:           defaultTopic,
-		Encoding:        defaultEncoding,
+		// 'otlp_spans' is being deprecated as the default topic.
```
Comment (pavolloffay) [inline exporter/kafkaexporter/factory.go:78]: we are not deprecating the topic name right? we want to use different topic names for different signals

Comment (pavolloffay) [reply exporter/kafkaexporter/factory.go:78]: @bogdandrutu this brings an interesting point. Maybe the factory should be capable of creating different defaults for different signals.

Comment (bogdandrutu) [reply exporter/kafkaexporter/factory.go:78]: @pavolloffay true, we should file an issue to discuss this.

File: CHANGELOG.md (hunk 8 above)
Comment (dshomoye) [reply CHANGELOG.md:26]: oh this is a typo from the last merge from master, fixing now

File: exporter/kafkaexporter/factory.go (hunk 11 above)
Comment (dshomoye) [reply exporter/kafkaexporter/factory.go:78]: I'll update the comment

Comment (pavolloffay) [reply exporter/kafkaexporter/factory.go]: @bogdandrutu created https://github.com/open-telemetry/opentelemetry-collector/issues/2160

**Comment by dshomoye** [comment]
Anything needed to merge this now?

**Comment by tigrannajaryan** [comment]
Merging this broke the build: https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector/4718/workflows/59a274c1-241d-4928-8411-2c5330c7fc47

I am going to revert to fix the build. @dshomoye after I revert please create a new PR with this changes. You will need to make some additional changes to make sure the build passes.
